src/content/en/updates/2019/02/trusted-types.md

```diff
@@ -0,0 +1,221 @@
+
+[Cross-Site Scripting](https://en.wikipedia.org/wiki/Cross-site_scripting) (XSS)
+is the most prevalent vulnerability affecting web applications. We see this
+reflected both in
+[our own data](https://security.googleblog.com/2016/09/reshaping-web-defenses-with-strict.html),
+and
+[throughout the industry](https://www.hackerone.com/sites/default/files/2017-06/The%20Hacker-Powered%20Security%20Report.pdf){: .external}.
+Practice shows that maintaining an XSS-free application is still a difficult
+challenge, especially if the application is complex. While solutions for
+preventing server-side XSS are well known,
+[DOM-based Cross-Site Scripting](https://www.owasp.org/index.php/DOM_Based_XSS){: .external}
+(DOM XSS) becomes a *growing* problem. For example, in
+[Google Vulnerability Reward Program](https://g.co/vrp) DOM XSS is already the
+most common variant.
+
+Why is that? We think it's caused by two separate issues:
+
+### XSS is easy to introduce
+
+DOM XSS occurs when one of injection sinks in DOM or other browser APIs is
+called with user-controlled data. For example, consider this snippet that
+intends to load a stylesheet for a given UI template the application uses:
+
+```js
+const templateId = location.hash.match(/tplid=([^;&]*)/)[1];
+// ...
+document.head.innerHTML += `<link rel="stylesheet" href="./templates/${templateId}/style.css">`
+```
+
+This code introduces DOM XSS by linking the attacker-controlled **source**
+(`location.hash`) with the injection **sink** (`innerHTML`). The attacker can
+exploit this bug by sending `https://example.com#tplid="><img src=x
+onerror=alert(1)>` URL for the victim to visit.
+
+It's easy to make this mistake in code, especially if the code changes often.
+For example, maybe `templateId` was once generated and validated on the server,
+so this value was trustworthy? When assigning to `innerHTML`', all we know is
+that the value is a string, but should it be trusted? Where does is come from
+really?
+
+Additionally, it's not just `innerHTML` that one must be careful of. In a
+typical browser environment, there are over 60 sink functions or properties that
+require this caution. The DOM API is **insecure by default** and requires
+special treatment to prevent XSS.
+
+### XSS is difficult to detect
+
+The code above is just an example, so it's trivial to see the bug. In practice,
+the sources and the sinks are often accessed in completely different application
+parts. The data from the source is passed around, and eventually reaches the
+sink. There are some functions that sanitize and verify the data. But was the
+right function called?
+
+Looking at the source code alone, it's difficult to know if it introduces a DOM
+XSS. It's not enough to grep the `.js` files for sensitive patterns. For one,
+the sensitive functions are used through various wrappers, and the actual
+vulnerability will look more like
+[this](https://hackerone.com/reports/158853){: .external}. But sometimes it's
+just impossible:
```

Comment:
Maybe start a new paragraph before "But"

Suggested rephrasing: "Sometimes it's not even possible to tell if a given piece of code is vulnerable by looking at it in isolation."

Reply:
The example is more about the inability to determine the presence of XSS statically (by human or a linter). I rephrased to 
"Sometimes it's not even possible to tell if a codebase is vulnerable by only
looking at it."